lib/ansible/galaxy/dependency_resolution/dataclasses.py

```diff
@@ -584,3 +584,15 @@ def __new__(cls, *args, **kwargs):
 
     def __init__(self, *args, **kwargs):
         super(Candidate, self).__init__()
+
+    @classmethod
+    def with_signatures_repopulated(cls, collection):
+        # type: (t.Type[Candidate], Candidate) -> Candidate
+        if collection.type != 'galaxy':
+            raise ValueError(
+                f"Invalid collection type for {collection}: unable to get signatures from a galaxy server."
```

Comment:
`TypeError` seems to be better-fitting semantically:
```suggestion
            raise TypeError(
                f"Invalid collection type ({collection.type!s}) for {collection!s}: "
                "unable to get signatures from a galaxy server."
```
(also, it might be useful to log the actual type)

Reply:
OTOH, this guard looks like an entry-contract check for a programming error that shouldn't be triggerable by user input so an `AssertionError` would be a natural fit:

```suggestion
            raise AnsibleAssertionError(
                f"Invalid collection type ({collection.type!s}) for {collection!s}: "
                "unable to get signatures from a galaxy server."
```

Reply:
I'm using the `__repr__` to display all attributes of the collection, not just the type so it can be identified easily.

I updated to it an AssertionError since this is just for developers to prevent accidentally calling for types that don't need signatures modified.